File: geode-core/src/main/java/org/apache/geode/cache/client/PoolManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.geode.cache.client;
import java.util.Map;

import org.apache.geode.cache.Region;
import org.apache.geode.internal.cache.PoolManagerImpl;
import org.apache.geode.distributed.DistributedSystem; // for javadocs

/**
 * Manages creation and access to {@link Pool connection pools} for clients.
 * <p>
 * To create a pool get a factory by calling {@link #createFactory}.
 * <p>
 * To find an existing pool by name call {@link #find(String)}.
 * <p>
 * To get rid of all created pool call {@link #close()}.
 *
 * @since GemFire 5.7
 *
 */
public final class PoolManager {
  
  private PoolManager() {
    // no instances allowed!
  }
  
  /**
   * Creates a new {@link PoolFactory pool factory},
   * which is used to configure and create new {@link Pool}s.
   * @return the new pool factory
   */
  public static PoolFactory createFactory() {
    return PoolManagerImpl.getPMI().createFactory();
  }

  /**
   * Find by name an existing connection pool returning
   * the existing pool or <code>null</code> if it does not exist.
   * @param name the name of the connection pool
   * @return the existing connection pool or <code>null</code> if it does not exist.
   */
  public static Pool find(String name) {
    return PoolManagerImpl.getPMI().find(name);
  }
  /**
   * Returns a map containing all the pools in this manager.
   * The keys are pool names
   * and the values are {@link Pool} instances.
   * <p> The map contains the pools that this manager knows of at the time of this call.
   * The map is free to be changed without affecting this manager.
   * @return a Map that is a snapshot of all the pools currently known to this manager.
   */
  public static Map<String,Pool> getAll() { 
    return PoolManagerImpl.getPMI().getMap();
  }
  /**
   * Unconditionally destroys all created pools that are in this manager.
   * @param keepAlive whether the server should keep the durable client's subscriptions alive for the <code>durable-client-timeout</code>.
   * @see DistributedSystem#connect for a description of <code>durable-client-timeout</code>.
   */
  public static void close(boolean keepAlive) {
    PoolManagerImpl.getPMI().close(keepAlive);
  }
  
  /**
   * Find the pool used by the given region.
   * @param region The region that is using the pool.
   * @return the pool used by that region or <code>null</code> if the region does
   * not have a pool. 
   */
  public static Pool find(Region<?,?> region) {
    return PoolManagerImpl.getPMI().find(region);
  }
  
  /**
   * Unconditionally destroys all created pools that are in this manager.
   */
  public static void close() {
    close(false);
  }
}

File: src/clicache/src/PoolManager.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#pragma once

//#include "gf_includes.hpp"
#include "Region.hpp"
#include "Pool.hpp"
#include "PoolManager.hpp"
#include "PoolFactory.hpp"
#include "CacheableString.hpp"
#include "impl/SafeConvert.hpp"

using namespace System;

namespace Apache
{
  namespace Geode
  {
    namespace Client
    {
namespace Generic
    {
      //generic<class TKey, class TValue>
      PoolFactory/*<TKey, TValue>*/^ PoolManager/*<TKey, TValue>*/::CreateFactory()
      {
        return PoolFactory/*<TKey, TValue>*/::Create(apache::geode::client::PoolManager::createFactory().ptr());
      }

      //generic<class TKey, class TValue>
      const Dictionary<String^, Pool/*<TKey, TValue>*/^>^ PoolManager/*<TKey, TValue>*/::GetAll()
      {
        apache::geode::client::HashMapOfPools pools = apache::geode::client::PoolManager::getAll();
        Dictionary<String^, Pool/*<TKey, TValue>*/^>^ result = gcnew Dictionary<String^, Pool/*<TKey, TValue>*/^>();
        for (apache::geode::client::HashMapOfPools::Iterator iter = pools.begin(); iter != pools.end(); ++iter)
        {
          String^ key = CacheableString::GetString(iter.first().ptr());
          Pool/*<TKey, TValue>*/^ val = Pool/*<TKey, TValue>*/::Create(iter.second().ptr());
          result->Add(key, val);
        }
        return result;
      }

      //generic<class TKey, class TValue>
      Pool/*<TKey, TValue>*/^ PoolManager/*<TKey, TValue>*/::Find(String^ name)
      {
        ManagedString mg_name( name );
        apache::geode::client::PoolPtr pool = apache::geode::client::PoolManager::find(mg_name.CharPtr);
        return Pool/*<TKey, TValue>*/::Create(pool.ptr());
      }

      //generic <class TKey, class TValue>
      Pool/*<TKey, TValue>*/^ PoolManager/*<TKey, TValue>*/::Find(Generic::Region<Object^, Object^>^ region)
      {
        //return Pool::Create(apache::geode::client::PoolManager::find(apache::geode::client::RegionPtr(GetNativePtr<apache::geode::client::Region>(region))).ptr());
        return Pool/*<TKey, TValue>*/::Create(apache::geode::client::PoolManager::find(apache::geode::client::RegionPtr(region->_NativePtr)).ptr());
      }

      //generic<class TKey, class TValue>
      void PoolManager/*<TKey, TValue>*/::Close(Boolean KeepAlive)
      {
        apache::geode::client::PoolManager::close(KeepAlive);
      }

      //generic<class TKey, class TValue>
      void PoolManager/*<TKey, TValue>*/::Close()
      {
        apache::geode::client::PoolManager::close();
      }
    }
  }
}

 } //namespace 
